truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLanguageService.java

```diff
@@ -152,7 +152,7 @@ private boolean validHostValue(Object hostValue, HostContext context) {
 
     @Override
     public boolean isHostValue(Object value) {
-        Object obj = HostLanguage.unwrapIfScoped(language, value);
+        Object obj = HostLanguage.unwrapIfScoped(language, HostLibrary.getUncached(), value);
```

Comment:
That is not true. Such a rule does not exist and does particularly not apply for most of the truffle.polyglot and truffle.host code. The only important thing is that the virtual call can be devirtualized reliably during PE otherwise it would cause a performance warning during compilation. This is generally true for all objects rooted in the PolyglotEngineImpl class like AbstractPolyglotHostService as the engine is always a constant in PE paths.

Reply:
Note that AbstractPolyglotHostService exists for polyglot isolates, as all host operations need to be handled differently there.

Reply:
The `@Override` claim, was just a (wrong) guess. I know PE doesn't like overwritten methods, but if it resolves to a single method, then it of course works fine.

To compensate for my erroneous judgment, I suggest following check: Modify the `unwrapIfScoped` method to reference some blacklisted method:
```diff
graal/vm$ git diff
diff --git a/truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLanguage.java b/truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLanguage.java
index 9706e5f5e56..a045c0ce7b0 100644
--- a/truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLanguage.java
+++ b/truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLanguage.java
@@ -61,6 +61,7 @@ import com.oracle.truffle.api.interop.UnsupportedMessageException;
 import com.oracle.truffle.api.nodes.Node;
 import com.oracle.truffle.api.nodes.RootNode;
 import com.oracle.truffle.host.HostMethodScope.ScopedObject;
+import java.io.File;
 
 /*
  * Java host language implementation.
@@ -94,6 +95,7 @@ final class HostLanguage extends TruffleLanguage<HostContext> {
     }
 
     static Object unwrapIfScoped(HostLanguage language, HostLibrary library, Object o) {
+        if (!new File(".").exists()) throw new IllegalStateException();
         HostInstance host = library.asHostInstance(o);
         if (language == null || !language.methodScopingEnabled) {
             return host;
```
and then compile `js` with native image. It gives seven violations like:
```
graal/vm$ mx --native-image=js --java-home ~/bin/labs-11/ --dy=/graal-js,/espresso --env ni-ce build
=== Found 7 compilation blocklist violations ===

Blocklisted method
  java.io.UnixFileSystem.getBooleanAttributes0(File)
called from
  java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:248)
  java.io.File.exists(File.java:831)
  com.oracle.truffle.host.HostLanguage.unwrapIfScoped(HostLanguage.java:98)
  com.oracle.truffle.host.HostObject.isInstance(HostObject.java:147) -> com.oracle.truffle.host.HostLanguageService.isHostObject(HostLanguageService.java:203)
  com.oracle.truffle.polyglot.EngineAccessor$EngineImpl.isHostObject(EngineAccessor.java:1080)
  com.oracle.truffle.api.TruffleLanguage$Env.isHostObject(TruffleLanguage.java:2116)
  com.oracle.truffle.js.builtins.ConstructorBuiltins$ConstructJavaImporterNode.constructJavaImporter(ConstructorBuiltins.java:2233)
  com.oracle.truffle.js.builtins.ConstructorBuiltinsFactory$ConstructJavaImporterNodeGen.execute(ConstructorBuiltinsFactory.java:4490)
  com.oracle.truffle.js.nodes.control.TopLevelAwaitModuleBodyNode$TopLevelAwaitModuleRootNode.execute(TopLevelAwaitModuleBodyNode.java:106)
  org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:650)
  org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:622)
 
```
All of them go thru `com.oracle.truffle.api.TruffleLanguage$Env.isHostObject`. I'll fix it.